### DISCUSSION TITLE
Issue while playing video in DIKSHA (Oracle platform)

### DISCUSSION THREAD

**Discussion by nileshranebr5941** [Sunbird-Ed/Community#642] [Bugs]
@gandham-santhosh 
@vinukumar-vs 
@Krishnaj20 
In DIKSHA, hosted on Oracle platform, we find an issue when playing a video file.
When the video is played via Oracle streaming service, we get an interruption when the pixels are changed.
The error in console is given below.  We have also observed that in SUnBird Azure env, a video plays perfectly all-right with no issues.
Requesting to throw some light on this issue

VIDEOJS: ERROR: (CODE:3 MEDIA_ERR_DECODE) video append of 3161488b failed for segment #0 in playlist 1-https://ctr-01273-ap-in.diksha-vod.qwilted-cds.cqloud.com/20180828/stream/origin/ocid1.streampackagingconfig.oc1.ap-mumbai-1.amaaaaaadut4dhyav7jewc3ojptsxvzy6b6wl6u75cp46q7ijqlslwxfuv2q/ocid1.mediaasset.oc1.ap-mumbai-1.amaaaaaadut4dhyai3zrf7ijas2qx6ri5242z6jc7hsbmnpawc7reeczgv2q/eyJhbGciOiJIUzI1NiJ9.eyJ2ZXIiOjE2NzYxNzE3ODg0MDcsImlzQXNzZXRQbGF5bGlzdCI6dHJ1ZSwiYXNzZXRQYXRoIjoiZGlrc2hhXzcyMF8yMzAwSy5tM3U4IiwicGFyZW50UGxheWxpc3RQYXRoIjoib3V0cHV0L2NvbnRlbnQvYXNzZXRzL2RvXzMxMzU5NDY1MDUyMjE5ODAxNjEyMDcxLzFfNF9lbmdfc2lnbmlmaWNhbmNlX2Vhcmx5X3llYXJzX2VkdWNhdGlvbi5tcDQvZGlrc2hhLW1hc3Rlci5tM3U4In0.S76gcqIv1GbAmAStVXNzZlenu4En8-jTPBhQqh8KqzU Tt {code: 3, message: 'video append of 3161488b failed for segment #0 in …U4In0.S76gcqIv1GbAmAStVXNzZlenu4En8-jTPBhQqh8KqzU'}

**Comment by gandham-santhosh** [comment]
@nileshranebr5941 from which service you are seeing this error? Was this working before and stopped working now?

@pallakartheekreddy @ddevadat please share your thoughts here

**Reply by heungheung** [reply]
@gandham-santhosh FYI
https://github.com/videojs/http-streaming/issues/1344

I think the player is using videojs, right?

**Reply by damithkothalawala** [reply]
@gandham-santhosh seems something wrong with videojs. As you can see on recording, it is taking long time to switch back and even getting stuck at some stages. 

Apart from that there were some deprecation warnings on console.


https://www.youtube.com/watch?v=Pbb3MhaO2tY

**Reply by gandham-santhosh** [reply]
Thanks @heungheung @damithkothalawala for the pointers. If some features are deprecated, should we need to upgrade the version or use different lib? @pallakartheekreddy @vinukumar-vs please chime in here

**Reply by damithkothalawala** [reply]
@gandham-santhosh I just did a sample setup with newer videojs over here https://apaccpt03.objectstorage.ap-mumbai-1.oci.customer-oci.com/p/Qz7Z7QIhQBymxM2OnymZL4N9sCCRYlzYjR5tOlBIBug0gSUJ3A22nAlWq45EuR1I/n/apaccpt03/b/videojs/o/player.html 

I still have the same issue with quality switching, meaning it is something to do with videojs as mentioned on the bug entry.

Meanwhile I am working on collecting more facts and doing a recording with hls players. Will comeback with facts later today.

**Reply by nileshranebr5941** [reply]
@gandham-santhosh    one question that I have is, this scenario works perfectly in https://staging.sunbirded.org/   which is hosted in Azure. Even when we change the pixels there the video plays smooth, there is not breaking in the video.
Could you just check, what version of the components are being used? Is any thing being done differently.?  Please advise

**Reply by heungheung** [reply]
what is the m3u8? are we talking about same problem of videojs?

**Comment by nileshranebr5941** [comment]
@gandham-santhosh   

The issue occurs for normal playing of the video.
For example when we search for a videos DO_ID, and then we click on play and then try to change pixel, this issue occurs.
Does not look like it was working earlier too.

**Reply by gandham-santhosh** [reply]
@nileshranebr5941 can you please look for error logs from the media streaming flink jobs and share here? For me it looks like some configurations need to be enabled at OCI media service end. @ddevadat request you to help here

**Reply by nileshranebr5941** [reply]
@gandham-santhosh 
also in network tab we get Cancelled for random streaming request, below is the streaming url.


https://ctr-01273-ap-in.diksha-vod.qwilted-cds.cqloud.com/20180828/stream/origin/ocid1.streampackagingconfig.oc1.ap-mumbai-1.amaaaaaadut4dhyav7jewc3ojptsxvzy6b6wl6u75cp46q7ijqlslwxfuv2q/ocid1.mediaasset.oc1.ap-mumbai-1.amaaaaaadut4dhyai3zrf7ijas2qx6ri5242z6jc7hsbmnpawc7reeczgv2q/eyJhbGciOiJIUzI1NiJ9.eyJ2ZXIiOjE2NzYxNzE3ODg0MDcsImlzQXNzZXRQbGF5bGlzdCI6dHJ1ZSwiYXNzZXRQYXRoIjoiZGlrc2hhXzQ4MF8xMjAwSy5tM3U4IiwicGFyZW50UGxheWxpc3RQYXRoIjoib3V0cHV0L2NvbnRlbnQvYXNzZXRzL2RvXzMxMzU5NDY1MDUyMjE5ODAxNjEyMDcxLzFfNF9lbmdfc2lnbmlmaWNhbmNlX2Vhcmx5X3llYXJzX2VkdWNhdGlvbi5tcDQvZGlrc2hhLW1hc3Rlci5tM3U4In0.2nZmTpOT5cORCt7SF0ztYa-VI7590rwgrz5lBw0_T7Y

**Reply by damithkothalawala** [reply]
@nileshranebr5941 please contact CDN Provider and try to get Origin logs after flushing cache for a unique resource id. Then only we can scope this down to a single issue.

As you know segments being cached by the CDN and it is responsible for the delivery. So as mentioned take all logs from CDN provider and let us know.

**Comment by nileshranebr5941** [comment]
@gandham-santhosh 

{this is what we get in console}
VIDEOJS: ERROR: (CODE:3 MEDIA_ERR_DECODE) video append of 3161488b failed for segment #0 in playlist 1-https://ctr-01273-ap-in.diksha-vod.qwilted-cds.cqloud.com/20180828/stream/origin/ocid1.streampackagingconfig.oc1.ap-mumbai-1.amaaaaaadut4dhyav7jewc3ojptsxvzy6b6wl6u75cp46q7ijqlslwxfuv2q/ocid1.mediaasset.oc1.ap-mumbai-1.amaaaaaadut4dhyai3zrf7ijas2qx6ri5242z6jc7hsbmnpawc7reeczgv2q/eyJhbGciOiJIUzI1NiJ9.eyJ2ZXIiOjE2NzYxNzE3ODg0MDcsImlzQXNzZXRQbGF5bGlzdCI6dHJ1ZSwiYXNzZXRQYXRoIjoiZGlrc2hhXzcyMF8yMzAwSy5tM3U4IiwicGFyZW50UGxheWxpc3RQYXRoIjoib3V0cHV0L2NvbnRlbnQvYXNzZXRzL2RvXzMxMzU5NDY1MDUyMjE5ODAxNjEyMDcxLzFfNF9lbmdfc2lnbmlmaWNhbmNlX2Vhcmx5X3llYXJzX2VkdWNhdGlvbi5tcDQvZGlrc2hhLW1hc3Rlci5tM3U4In0.S76gcqIv1GbAmAStVXNzZlenu4En8-jTPBhQqh8KqzU Tt {code: 3, message: 'video append of 3161488b failed for segment #0 in …U4In0.S76gcqIv1GbAmAStVXNzZlenu4En8-jTPBhQqh8KqzU'}
(anonymous) @ scripts.5f77aaea866551ed58c6.js:17
r.error @ scripts.5f77aaea866551ed58c6.js:17
r.error @ scripts.5f77aaea866551ed58c6.js:29
(anonymous) @ scripts.5f77aaea866551ed58c6.js:33
Ue.n.dispatcher.n.dispatcher @ scripts.5f77aaea866551ed58c6.js:17
He @ scripts.5f77aaea866551ed58c6.js:17
Ye.trigger @ scripts.5f77aaea866551ed58c6.js:17
(anonymous) @ scripts.5f77aaea866551ed58c6.js:33
Ue.n.dispatcher.n.dispatcher @ scripts.5f77aaea866551ed58c6.js:17
He @ scripts.5f77aaea866551ed58c6.js:17
Ye.trigger @ scripts.5f77aaea866551ed58c6.js:17
i.handleAppendError_ @ scripts.5f77aaea866551ed58c6.js:30
(anonymous) @ scripts.5f77aaea866551ed58c6.js:30
e.invokeTask @ polyfills.cb883d48238530c0217a.js:13
onInvokeTask @ main.9c9d17935c808d1b52f1.js:1250
e.invokeTask @ polyfills.cb883d48238530c0217a.js:13
t.runTask @ polyfills.cb883d48238530c0217a.js:13
t.invokeTask @ polyfills.cb883d48238530c0217a.js:13
_ @ polyfills.cb883d48238530c0217a.js:44
b @ polyfills.cb883d48238530c0217a.js:44

**Reply by gandham-santhosh** [reply]
We tried with one video content and tried to change the quality and it worked fine https://diksha.gov.in/play/content/do_31332854545122918416252?contentType=Resource

I also noticed that if you change the setting multiple times without giving any time gap, it is throwing the error. Are you also seeing the same behaviour?

**Reply by nileshranebr5941** [reply]
@gandham-santhosh 
The first image is showing the pixel option, the moment the pixel is changed, the error comes, that is available in second screen shot.
also in which PODS i can find the relevant logs.


<img width="863" alt="image" src="https://github.com/Sunbird-Ed/Community/assets/134708560/a033de68-b004-4313-a941-18920de484b8">


<img width="934" alt="image" src="https://github.com/Sunbird-Ed/Community/assets/134708560/74f7d2ba-48ee-466d-b519-b2926638909c">

**Reply by vinukumar-vs** [reply]
@nileshranebr5941 Can you verify the streaming URL generated for the content (you can find in the read API response) is able to play in the below sample demo application of VideoJS.

<img width="1532" alt="image" src="https://github.com/Sunbird-Ed/Community/assets/4496096/ae3b4109-1f0d-4e08-a2ba-14480dca4bca">


[VideoJS Streaming URL demo ](https://videojs-http-streaming.netlify.app/?debug=false&autoplay=false&muted=true&fluid=false&minified=false&sync-workers=false&liveui=true&llhls=true&url=https%3A%2F%2Fmibm2gd4gvfa5jjuaizursbvvi.apigateway.ap-mumbai-1.oci.customer-oci.com%2Focid1.mediaasset.oc1.ap-mumbai-1.amaaaaaadut4dhyayh25vm57544mxwncszrsnovmth5nrubcsaokoc4yfeka%2Fmaster.m3u8&type=application%2Fx-mpegURL&keysystems=&buffer-water=false&exact-manifest-timings=false&pixel-diff-selector=false&network-info=false&dts-offset=false&offset-each-segment=false&override-native=true&preload=auto&mirror-source=true&forced-subtitles=false)

User the Chrome Network tab `Throttling` option to change the network speed. After changing the throttling, if the video is playing fine means, streaming server is able to serve the data properly & videoJS library is able to stream the videos as per network speed.

In case, if the video is not playing for the specific network speed(throttling) there can be 2 reasons
1. Verify the video streaming logs to check is there any issue in streaming the video content.
2. VideoJS library is not able to recognise the streaming URL's for the specific network speed(throttling). This will be challenging to find what is the problem in VideoJS library to stream.

I would suggest to verify the same with different streamingURL's(Same CSP or other CSP) to find out the root cause of the issue.

@damithkothalawala @gandham-santhosh @heungheung @Krishnaj20 @pallakartheekreddy @ddevadat

**Comment by nileshranebr5941** [comment]
@gandham-santhosh 

do_31359465052219801612071, use this do id and when its playing change the pixel.
Though the error is intermittent, it comes often

**Comment by damithkothalawala** [comment]
Hi @gandham-santhosh  . Can you please help me with getting the default request timeout value for player's playlist load and segment loading ?. 
Thank You

**Reply by gandham-santhosh** [reply]
@pallakartheekreddy @vinukumar-vs  request you to check and share this

**Reply by HarishGangula** [reply]
@damithkothalawala 

Video.js does not have a built-in default timeout value for playlist and segment loading. The default timeout behavior is determined by the browser's network timeout settings, which can vary between different browsers.

Most modern web browsers have default timeout values for network requests, and they typically range from 30 seconds to 2 minutes. If a request (e.g., for a video segment or playlist) takes longer than the browser's timeout, it may result in a timeout error.

**Reply by damithkothalawala** [reply]
my observation was that for example, when you play a content like https://diksha.gov.in/play/content/do_3130921029093867521984?contentType=ExplanationResource 

1. Let it start to play
2. Switch quality . Then you will see NS_BINDING_ABORTED seems it is to stop loading next segment 
3. When you switch too soon, then you will see Unable to load content error mentioned by @nileshranebr5941 

reason I can think of is that 

1. Segment size is around 5-6s and will take sometime to load HQ segments
2. If switched before buffering, with abort (for currently loading segment) it throws corrupted segment error.

So as the conclusion for my findings, this random error happens when

1. when observed high latency to CDN
2. when quality switched too often (related to videoJs bug)

hope this is helpful.
